crates/wasm-mutate/src/lib.rs

```diff
@@ -9,13 +9,60 @@
 
 #![cfg_attr(not(feature = "structopt"), deny(missing_docs))]
 
-use std::sync::Arc;
+use std::{io::Write, sync::Arc};
 
+use mutators::{Mutator, ReturnI32SnipMutator, SetFunction2Unreachable, RemoveExportMutator, NoMutator};
+use rand::{rngs::StdRng, Rng, SeedableRng};
 #[cfg(feature = "structopt")]
 use structopt::StructOpt;
 
 mod error;
+
+pub mod mutators;
+
 pub use error::{Error, Result};
+use wasm_encoder::{CodeSection, Module, SectionId};
+use wasmparser::{Chunk, Parser, Payload};
+
+
+macro_rules! mutators {
+    (
+        $(
+            ($tpe: pat$(, $mutation:expr)*),
+        )*
+    ) => {
+        fn mutate<'a>(p:&'a mut Payload<'a>, seed: &mut StdRng, context: &'a WasmMutate, chunk:&'a [u8], out_buffer: &'a mut dyn Write) -> ()
+        {
+            match  p {
+                $(
+                    $tpe => {
+                        let options = [
+                            #[cfg(feature="no-mutator")]
+                                move | config, chunk, out, payload  | (Some(String::from("NoMutator")), NoMutator{}.mutate(config, chunk, out, payload)),
+                            $(  
+                                move | config, chunk, out, payload  | (Some($mutation.name()), $mutation.mutate(config, chunk, out, payload)), 
+                            )*
+                        ];// Instantiate after select
+
+                        
+                        // TODO this is maybe not the best way
+                        let mutation = options.get(seed.gen_range(0, options.len())).unwrap();
+                        let (name, _ ) = mutation(context, chunk, out_buffer, p);
```

Comment:
Thats right...one option to construct the possible mutators is to do this mutator's chain "construction" in two passes, the first one to evaluate and define the valid mutators and the last phase to apply them. In terms of performance this could be problematic since we do a O(2N) process, but we can connect somekind of cache to determine the valid mutators right away if the same module is passed again to the tool. WDYT?

Reply:
I think doing two passes is fine. I think we can do something similar to what Wizer does:

* Parse the Wasm into a very lightweight AST, that is essentially `Vec<wasm_encoder::RawSection>` plus a little bit of meta info (`has_memory: bool` / `has_table: bool` / `has_function: bool` / etc...)

  * For reference, here is where Wizer parses the Wasm into its AST (which is a bit more involved than what I sketched above, but not too much more involved): https://github.com/bytecodealliance/wizer/blob/main/src/parse.rs#L17

* Then build our list of applicable mutations, where each mutator can use that meta info from the AST to quickly determine whether they are applicable to this Wasm module or not

* Run the mutator, simultaneously encoding a new, mutated Wasm module

Reply:
Gotcha !, lets do that then